### PR TITLE
Enotice fix on userDashboard with contributions, replace deprecated functions with api4v calls

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Enotice fix on userDashboard with contributions

- occurs when debugging is enabled & contact dashboard is viewed for a contact who is the target of soft credits

Before
----------------------------------------
image

After
----------------------------------------
Notice cleared

image

Technical Details
----------------------------------------
I wound up with some scope creep here as the notice turned out to be in the soft credits secion not the main section. This turned out to be accessing a function that was already recommended for deprecation & only accessed by tests. I switched to moving the test to check the dashboard rendered rather than the deprecated function.

As I was wrapping up discovered the `ContributionRecur` section used 3 different forms of deprecated function - including a `require_once` on the api utils file and the potentially bug-inducing 

`   $recurStatus = CRM_Contribute_PseudoConstant::contributionStatus(NULL, 'label');` 

accessing the wrong option group....

So in the end as I was deep in adding the tests I fixed all except one of the deprecated function uses

Comments
----------------------------------------
The `tpl` variables are also changed here a little - mostly around the `honorRows` which is not that accurate & is renamed to `soft_credit_contributions` & standardised with the contributions array. This should be surfaced in the release notes

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
CRM_Contact_Page_View_UserDashBoardTest::testDashboardContentContributions
Undefined array key "id"

/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/default/files/civicrm/templates_c/en_US/%%D5/D5B/D5B2D140%%UserDashboard.tpl.php:35
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1914
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/default/files/civicrm/templates_c/en_US/%%5B/5B5/5B5E9910%%UserDashBoard.tpl.php:31
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1914
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/default/files/civicrm/templates_c/en_US/%%F7/F77/F77C7890%%CMSPrint.tpl.php:67
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1914
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/default/files/civicrm/templates_c/en_US/%%85/854/854F26FC%%unittests.tpl.php:6
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/packages/Smarty/Smarty.class.php:1273
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/CRM/Core/Smarty.php:189
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/CRM/Core/Page.php:254
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/CRM/Contact/Page/View/UserDashBoard.php:195
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contact/Page/View/UserDashBoardTest.php:259
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/tests/phpunit/CRM/Contact/Page/View/UserDashBoardTest.php:159
/home/jenkins/bknix-dfl/build/core-24861-1vbop/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:265
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:1721

Passing locally & I can't figure out what line it is :-( (smarty compiled is scrunched)

**Comment by yashodha** [comment]
@eileenmcnaughton in _Your Contributions_ section, the contributions are still not sorted by received date descending (look at the after screenshot)

image

You might to want fix that as well while you are at it.

**Comment by colemanw** [comment]
General note, just to maks sure this is on your radar:
The biggest difference between using the DAO vs the API to display data on a page is that the DAO would fetch strings that are pre-escaped (`<` and `>` characters are html-escaped in the db) and the API does not.

**Comment by eileenmcnaughton** [comment]
@yashodha - @colemanw  & @seamuslee001 agreed your search order change made sense so I've included it. I had to adjust a test as a result

@colemanw I added escaping.

File: CRM/Contribute/Page/UserDashboard.php [hunk 1]
```diff
@@ -18,112 +21,84 @@ class CRM_Contribute_Page_UserDashboard extends CRM_Contact_Page_View_UserDashBo
...
             'id' => Civi::settings()->get('default_invoice_page'),
-            'ccid' => $row['contribution_id'],
+            'ccid' => $row['id'],
             'cs' => $this->getUserChecksum(),
             'cid' => $row['contact_id'],
           ]),
         ];
       }
     }
+    unset($row);
```
Comment (seamuslee001) [inline CRM/Contribute/Page/UserDashboard.php:49]: @eileenmcnaughton why is this here?

File: CRM/Contribute/Page/UserDashboard.php [hunk 2]
```diff
@@ -163,11 +133,68 @@ public function isIncludeInvoiceLinks() {
...
+        'contact_id',
+        'contact_id.display_name',
+        'contribution_status_id:name',
+        'contribution_status_id:label',
+        'financial_type_id:label',
+        'currency',
+        'amount_level',
+        'contact_id,',
+        'source',
+        'balance_amount',
```
Comment (seamuslee001) [inline CRM/Contribute/Page/UserDashboard.php:174]: Just noting this is ok because it is a computed field https://github.com/civicrm/civicrm-core/blob/5d378881aa065a348cc2952c5393814a82e38474/Civi/Api4/Service/Spec/Provider/ContributionGetSpecProvider.php#L50

File: CRM/Contribute/Page/UserDashboard.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Contribute/Page/UserDashboard.php:49]: it's just for clarity when you have a variable bleed out of a loop & it is a reference there is a bit of a risk of doing something odd with it

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
@colemanw @yashodha this is passing now - I think all comments are addressed

**Comment by yashodha** [comment]
@eileenmcnaughton you might want to squash the commits

**Comment by eileenmcnaughton** [comment]
@yashodha done

**Comment by yashodha** [comment]
@eileenmcnaughton merging this PR

**Comment by eileenmcnaughton** [comment]
Yay - thanks @yashodha
